Performance improvements related to map::process_fields

Aivean opened the pull request:
#### Summary

SUMMARY: Performance "Performance improvements related to map::process_fields"

#### Purpose of change

Continuation of the work started in #43469.
For "fast" simulation, i.e. when reading or crafting, `map::process_fields` takes ≈70% of the turn time.

#### Describe the solution

Several things were improved that effectively cut `map::process_fields` time in half:

1. Fixed [a bug](https://github.com/CleverRaven/Cataclysm-DDA/compare/master...Aivean:fields-performance-improvements?expand=1#diff-6065ddfbb156a26f24f995b70401a596R571) that was triggering wandering fields calculation for every tile
2. [here](https://github.com/CleverRaven/Cataclysm-DDA/compare/master...Aivean:fields-performance-improvements?expand=1#diff-ab005ad9757130721ffb66a09833090aR5539) `int_id` was converted to `string_id` unnecessarily
3. Extracted `field_id` creation [outside of the loop](https://github.com/CleverRaven/Cataclysm-DDA/compare/master...Aivean:fields-performance-improvements?expand=1#diff-6065ddfbb156a26f24f995b70401a596R399)
4. minor refactoring in `field::add_field` and `field::remove_field` which has negligible performance impact


#### Describe alternatives you've considered

I was also looking closely at `field::find_field`. Currently after all improvements to `process_fields` it takes as much as 8% of `game::do_turn` ([screenshot](https://user-images.githubusercontent.com/2865203/92344859-6c314280-f07c-11ea-868f-ddc580bee71c.png)). Feels like `std::map` is not the best fit for `field::_field_type_list`, as majority of the time it has 0-1 elements. I tried `unordered_map`, but there was no difference in performance. It would be interesting to try `std::vector<std:pair<field_type_id, field_entry>>` or maybe even handcrafted biased data structure.

#### Testing

Profiling for performance evaluation.
For correctness, I manually ensured that game loads, fields and wandering fields still work as expected.


#### Additional context
Both profiling sessions are from 30 seconds of reading in the lab, same character, same place, same everything, different builds.

Before:
![image](https://user-images.githubusercontent.com/2865203/92343809-933a4500-f079-11ea-8dc8-07bef9371b8b.png)

After:
![image](https://user-images.githubusercontent.com/2865203/92343870-c41a7a00-f079-11ea-859d-680acd765d31.png)


Note, this is a sampling profiling from essentially a busy loop, so if you see unchanged methods taking more time (like `game::monmove`) it means that because game turn was taking less absolute time, these methods were executed more times during the 30s session. What is relevant is relative % of `map::process_fields`.

Aivean commented:
@kevingranade , this is performance related, you might be interested.

AMurkin commented on src/map_field.cpp:
```diff
@@ -396,6 +396,21 @@ If you need to insert a new field behavior per unit time add a case statement in
 bool map::process_fields_in_submap( submap *const current_submap,
                                     const tripoint &submap )
 {
+    // create all ids ids once before the loop
```
```suggestion
    // create all ids once before the loop
```

ZhilkinSerg commented on src/field.cpp:
```diff
@@ -190,14 +190,15 @@ bool field::add_field( const field_type_id &field_type_to_add, const int new_int
-    if( field_type_to_add.obj().priority >= _displayed_field_type.obj().priority ) {
-        _displayed_field_type = field_type_to_add;
-    }
+
```
```suggestion
```

ZhilkinSerg commented on src/field.cpp:
```diff
@@ -215,10 +216,10 @@ bool field::remove_field( const field_type_id &field_to_remove )
-    } else {
-        _displayed_field_type = field_type_id( "fd_null" );
+    _displayed_field_type = fd_null;
+
```
```suggestion
```

ZhilkinSerg commented on src/map.cpp:
```diff
@@ -4195,7 +4195,7 @@ map_stack::iterator map::i_rem( const tripoint &p, const map_stack::const_iterat
         debugmsg( "Tried to remove items at (%d,%d) but the submap is not loaded", l.x, l.y );
         nulitems.clear();
-        return map_stack{ &nulitems, p, this }.begin();
+        return map_stack{ &nulitems, p, this } .begin();
```
```suggestion
        return map_stack{ &nulitems, p, this }.begin();
```

Aivean replied:
Weirdly enough, this space before `.begin` was put there by astyle.

Aivean commented:
@ZhilkinSerg , seems like your code suggestions were marked as outdated. Can you please unmark them if they are still relevant?

jbytheway commented:
Agreed that this is an example of something that would be good as a `flat_map` (which is essentially a map implemented as `std::vector<std:pair<field_type_id, field_entry>>`.  I have an old in-progress branch where I started implementing a `flat_map` for CDDA, but for annoying technical reasons it's hard to implement without undefined behaviour.  Regardless, I think that's best put off for a separate PR.

jbytheway commented on src/map_field.cpp:
```diff
@@ -396,6 +396,21 @@ If you need to insert a new field behavior per unit time add a case statement in
                                     const tripoint &submap )
 {
+    // create all ids once before the loop
+    const field_type_id fd_acid( "fd_acid" );
+    const field_type_id fd_extinguisher( "fd_extinguisher" );
+    const field_type_id fd_fire( "fd_fire" );
+    const field_type_id fd_fungal_haze( "fd_fungal_haze" );
+    const field_type_id fd_fire_vent( "fd_fire_vent" );
+    const field_type_id fd_flame_burst( "fd_flame_burst" );
+    const field_type_id fd_electricity( "fd_electricity" );
+    const field_type_id fd_push_items( "fd_push_items" );
+    const field_type_id fd_shock_vent( "fd_shock_vent" );
+    const field_type_id fd_acid_vent( "fd_acid_vent" );
+    const field_type_id fd_bees( "fd_bees" );
+    const field_type_id fd_incendiary( "fd_incendiary" );
+    const field_type_id fd_fungicidal_gas( "fd_fungicidal_gas" );
```
The inner loop here is mostly a case of "do a different thing for each field type".  This could be refactored as a `std::unordered_map<field_type_id, std::function<void()>>` initialized with a bunch of functions, one for each `if` block in the inner loop.  I think you could get away with storing such a map as a `static const` value so it would only need to be constructed once.  (As discussed in other PRs this would be bad if the `int_id`s for these fields changed without restarting CDDA, but I *think* it should be OK for these since they're all core fields; maybe it's a bit too fragile, though...).

It would be safer to use a `std::unordered_map<field_type_str_id, std::function<void()>>`; you'd then have to pay for one `int_id` -> `string_id` conversion per field (which is pretty cheap; just a function call or two and an array lookup) and a string hash conputation.

(You probably actually need to pass some arguments to the functions stored in the map, so it wouldn't actually be a `std::function<void()>`; also you could use a function pointer type instead of `std::function` since I don't think we'd want to use closures here).

Anyway, this is all just brainstorming.  The PR as it stands is already a good step in the right direction.

Zireael07 replied:
Good point, but I think it's something to do in a future PR, as this one is strictly about performance?

Aivean replied:
I thought something along the lines.  However, I decided to leave it as is.  From the performance standpoint, handful of ifs have negligible performance impact (I think it might be comparable with a single string hash calculation). Your suggestion might be more readable in a long run, especially if the number of cases grows.

BTW, field ids cannot be static, I tried that and game crashes on start.

Aivean replied:
If there are no objections, I'll leave this suggestion for the future research.

jbytheway commented on src/map_field.cpp:
```diff
@@ -430,18 +445,20 @@ bool map::process_fields_in_submap( submap *const current_submap,
                 // The field might have been killed by processing a neighbor field
                 if( !cur.is_field_alive() ) {
-                    if( !cur.get_field_type().obj().get_transparent( cur.get_field_intensity() - 1 ) ) {
+                    if( !curtype.obj().get_transparent( cur.get_field_intensity() - 1 ) ) {
```
```suggestion
                    if( !curtype->get_transparent( cur.get_field_intensity() - 1 ) ) {
```

jbytheway commented on src/map_field.cpp:
```diff
@@ -452,9 +469,11 @@ bool map::process_fields_in_submap( submap *const current_submap,
+                    curtype = fd_null;
                 }
 
+                const field_type &curtype_obj = curtype.obj();
```
```suggestion
                const field_type &curtype_obj = *curtype;
```
This is up to personal taste.

jbytheway commented on src/map_field.cpp:
```diff
@@ -430,18 +445,20 @@ bool map::process_fields_in_submap( submap *const current_submap,
                 field_entry &cur = it->second;
+
+                // Holds cur.get_field_type() as that is what the old system used before rewrite.
+                field_type_id curtype = cur.get_field_type();
```
FWIW I'd rename `curtype` to `curtype_id` and `curtype_obj` to `curtype`.

Aivean commented:
@jbytheway , I committed the suggested changes.

In addition, I slightly altered the semantics behind the `field::_displayed_field_type`. I changed `field_add` to maintain the invariant that `field` cannot store the `fd_null` field type id. Which means that if `_displayed_field_type == fd_null` the `field` is empty. So we have a quick way to avoid iterations over empty maps altogether.  I added [the same shortcut to the `find_field`](https://github.com/CleverRaven/Cataclysm-DDA/pull/44013/commits/212055197f7fcc1520ca9c2a9a5bfc32998ede58#diff-4ad10df75ab302c4b9ae78f758a32003R160). 

As a pleasant side effect, now fields with priority<0 are correctly displayed (before they were overridden by `fd_null` with priority=0).

However, this optimization does not help much in the heavy `spread_gas` scenarios, as gasses tend to spread around fields of the same type (so quick shortcut in `find_field` does nothing).

kevingranade commented:
Just FYI, what this needs in the future is for fields to decide on creation when they will next have a state change as per https://en.wikipedia.org/wiki/Gillespie_algorithm
This reduces the periodic computations to JUST the ticks when a state transition occurs instead of sweeping over all active fields each turn.
